### PR TITLE
Fix one of DCS constructors that cause NRE in wcf

### LINKED ISSUE #2435: DataContractSerializer constructor causes NullRefException when serializing object
When using [this](https://github.com/dotnet/corefx/blob/6e5d09b0334ca8735962f29f38a390c1298639c0/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs#L51) constructor for the DataContractSerializer:
```csharp
public DataContractSerializer(Type type, string rootName, string rootNamespace, IEnumerable<Type> knownTypes)
```
The resulting serializer can throw a `NullReferenceException`. This is caused by the creation of `XmlDictionaryString` objects which have a `null` `XmlDictionary`. 
There's also a bug/change in behavior in `XmlDictionaryString` compared with Desktop. On Desktop, if you specify a `null` value for the dictionary when constructing a new instance, an `ArgumentNullException` will be thrown. In corefx, it will attempt to create an empty instance of `IXmlDictionary` and assign that. It then straight away assigns the passed in `null` value to the dictionary effectively throwing away the empty dictionary that was created.  
Some of the Xml writers require the dictionary for an `XmlDicationary` string to be non-null. This is causing a `NullRefException` to be thrown. Here is a partial call stack from WCF:  

```csharp
   at System.Xml.XmlBinaryWriterSession.PriorityDictionary`2.TryGetValue(K key, V& value)
   at System.Xml.XmlBinaryWriterSession.TryLookup(XmlDictionaryString s, Int32& key)
   at System.Xml.XmlBinaryNodeWriter.TryGetKey(XmlDictionaryString s, Int32& key)
   at System.Xml.XmlBinaryNodeWriter.WriteXmlnsAttribute(String prefix, XmlDictionaryString ns)
   at System.Xml.XmlBaseWriter.NamespaceManager.DeclareNamespaces(XmlNodeWriter writer)
   at System.Xml.XmlBaseWriter.EndStartElement()
   at System.Xml.XmlBaseWriter.AutoComplete(WriteState writeState)
   at System.Xml.XmlBaseWriter.StartElement(String& prefix, String localName, String ns, XmlDictionaryString xNs)
   at System.Xml.XmlBaseWriter.WriteStartElement(String prefix, XmlDictionaryString localName, XmlDictionaryString namespaceUri)
   at System.Runtime.Serialization.XmlWriterDelegator.WriteStartElementPrimitive(XmlDictionaryString localName, XmlDictionaryString namespaceUri)
   at System.Runtime.Serialization.XmlObjectSerializerWriteContext.WriteString(XmlWriterDelegator xmlWriter, String value, XmlDictionaryString name, XmlDictionaryString ns)
   at WriteFaultDetailToXml(XmlWriterDelegator , Object , XmlObjectSerializerWriteContext , ClassDataContract )
   at System.Runtime.Serialization.ClassDataContract.WriteXmlValue(XmlWriterDelegator xmlWriter, Object obj, XmlObjectSerializerWriteContext context)
   at System.Runtime.Serialization.XmlObjectSerializerWriteContext.WriteDataContractValue(DataContract dataContract, XmlWriterDelegator xmlWriter, Object obj, RuntimeTypeHandle declaredTypeHandle)
   at System.Runtime.Serialization.XmlObjectSerializerWriteContext.SerializeWithoutXsiType(DataContract dataContract, XmlWriterDelegator xmlWriter, Object obj, RuntimeTypeHandle declaredTypeHandle)
   at System.Runtime.Serialization.DataContractSerializer.InternalWriteObjectContent(XmlWriterDelegator writer, Object graph, DataContractResolver dataContractResolver)
   at System.Runtime.Serialization.DataContractSerializer.InternalWriteObject(XmlWriterDelegator writer, Object graph, DataContractResolver dataContractResolver)
   at System.Runtime.Serialization.XmlObjectSerializer.WriteObjectHandleExceptions(XmlWriterDelegator writer, Object graph, DataContractResolver dataContractResolver)
   at System.Runtime.Serialization.XmlObjectSerializer.WriteObjectHandleExceptions(XmlWriterDelegator writer, Object graph)
   at System.Runtime.Serialization.XmlObjectSerializer.WriteObject(XmlDictionaryWriter writer, Object graph)
```

### PR DESCRIPTION
**Pull request by khdang**
Verified that this fixes #2435 by aligning with the behavior on Desktop

cc: @mconnew

### CODE REVIEW

**Comment by Maxwe11** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
Looks fine to me, but I don't know enough about this scenarios to sign-off.  @mconnew?

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryString.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ public class XmlDictionaryString
         public XmlDictionaryString(IXmlDictionary dictionary, string value, int key)
         {
             if (dictionary == null)
-                _dictionary = new EmptyStringDictionary();
+                throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ArgumentNullException("dictionary"));
```
Comment (krwq) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryString.cs]: This seems like a breaking change although according to MSDN we should be throwing ArgumentNullException. I can also see that we do exactly the same thing (new behavior) on desktop:
http://referencesource.microsoft.com/#System.Runtime.Serialization/System/Xml/XmlDictionaryString.cs,6fc6fdc6bb5abc25

Was it some fairly new fix which we forgot to port to corefx?

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs [hunk 2]
```diff
@@ -50,7 +50,8 @@ public DataContractSerializer(Type type, string rootName, string rootNamespace)
 
         public DataContractSerializer(Type type, string rootName, string rootNamespace, IEnumerable<Type> knownTypes)
         {
-            Initialize(type, new XmlDictionaryString(null, rootName, 0), new XmlDictionaryString(null, rootNamespace, 0), knownTypes, int.MaxValue, false, false, null, false);
+            XmlDictionary dictionary = new XmlDictionary(2);
+            Initialize(type, dictionary.Add(rootName), dictionary.Add(DataContract.GetNamespace(rootNamespace)), knownTypes, int.MaxValue, false, false, null, false);
```
Comment (krwq) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs]: Should we do similar fix below for the case where both NET_NATIVE and MERGE_DCJS are not defined?

**Comment by krwq** [comment]
LGTM

**Comment by stephentoub** [comment]
@khdang, is this still in process?  @krwq had a couple of questions, and I'm wondering if they're resolved?

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryString.cs (hunk 1 above)
Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryString.cs]: As far as I understand we want to have this to have the same behavior on desktop but I'm not aware of any related fix.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs (hunk 2 above)
Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs]: This fix is for this particular constructor which is used regardless of NET_NATIVE and MERGE_DCJS.

**Comment by khdang** [comment]
Yes it is in process. I've replied to @krwq questions.

**Comment by stephentoub** [comment]
Thanks.

**Pull request merged**